dbms/src/Storages/StorageHDFS.cpp

```diff
@@ -0,0 +1,178 @@
+    Context & context_)
+    : IStorage(columns_), uri(uri_), format_name(format_name_), table_name(table_name_), context_global(context_)
+{
+}
+
+namespace
+{
+    class StorageHDFSBlockInputStream : public IProfilingBlockInputStream
+    {
+    public:
+        StorageHDFSBlockInputStream(const String & uri,
+            const String & format,
+            const String & name_,
+            const Block & sample_block,
+            const Context & context,
+            size_t max_block_size)
+            : name(name_)
+        {
+            // Assume no query and fragment in uri, todo, add sanity check
+            String fuzzyFileNames;
+            String uriPrefix = uri.substr(0, uri.find_last_of('/'));
+            if (uriPrefix.length() == uri.length())
+            {
+                fuzzyFileNames = uri;
+                uriPrefix.clear();
+            }
+            else
+            {
+                uriPrefix += "/";
+                fuzzyFileNames = uri.substr(uriPrefix.length());
+            }
+
+            std::vector<String> fuzzyNameList = parseDescription(fuzzyFileNames, 0, fuzzyFileNames.length(), ',' , 100/* hard coded max files */);
+
+            // Don't support | for globs compatible
+            //std::vector<std::vector<String> > fileNames;
+            //for(auto fuzzyName : fuzzyNameList)
+            //    fileNames.push_back(parseDescription(fuzzyName, 0, fuzzyName.length(), '|', 100));
+
+            BlockInputStreams inputs;
+
+            //for (auto & vecNames : fileNames)
+            for (auto & name: fuzzyNameList)
+            {
+                std::unique_ptr<ReadBuffer> read_buf = std::make_unique<ReadBufferFromHDFS>(uriPrefix + name);
+
+                inputs.emplace_back(
+                    std::make_shared<OwningBlockInputStream<ReadBuffer>>(
+                        FormatFactory::instance().getInput(format, *read_buf, sample_block, context, max_block_size),
+                        std::move(read_buf)));
+            }
+
+            if (inputs.size() == 0)
+                throw Exception("StorageHDFS inputs interpreter error", ErrorCodes::BAD_ARGUMENTS);
+
+            if (inputs.size() == 1)
+            {
+                reader = inputs[0];
+            }
+            else
+            {
+                reader = std::make_shared<UnionBlockInputStream<> >(inputs, nullptr, context.getSettingsRef().max_distributed_connections);
+            }
+        }
+
+        String getName() const override
+        {
+            return name;
+        }
+
+        Block readImpl() override
+        {
+            return reader->read();
+        }
+
+        Block getHeader() const override
+        {
+            return reader->getHeader();
+        }
+
+        void readPrefixImpl() override
+        {
+            reader->readPrefix();
+        }
+
+        void readSuffixImpl() override
+        {
+            auto explicitReader = dynamic_cast<UnionBlockInputStream<> *>(reader.get());
```

Comment:
This is uncompilable code. You have to remove empty diamond `<>`.

Reply:
Fixed. UnionBlockInputStream was templated, but it is not after rebasing code.